File: reference/elixir/ampersand_core/lib/ampersand_core/compose.ex
defmodule AmpersandCore.Compose do
  @moduledoc """
  Capability composition utilities for the minimal [&] protocol reference
  implementation.

  This module models capability declarations as set-like maps keyed by protocol
  capability identifiers such as `&memory.graph` or `&reason.argument`.

  The composition rules are intentionally simple:

  - capability identifiers are normalized into a sorted list for comparison
  - composing disjoint capability sets merges them
  - composing identical bindings is idempotent
  - composing conflicting bindings for the same capability returns an error

  These rules give the implementation practical ACI-style behavior:

  - commutative for compatible declarations
  - associative for compatible declarations
  - idempotent for repeated declarations
  - identity via the empty capability set
  """

  @type capability_id :: String.t()
  @type capability_binding :: map()
  @type capability_map :: %{optional(capability_id()) => capability_binding()}
  @type input :: capability_map() | map()
  @type compose_error ::
          {:conflicting_binding, capability_id(), capability_binding(), capability_binding()}
          | {:invalid_input, term()}
  @type compose_result :: {:ok, capability_map()} | {:error, compose_error()}

  @identity %{}

  @doc """
  Returns the identity value for composition.
  """
  @spec identity() :: capability_map()
  def identity, do: @identity

  @doc """
  Returns a sorted, normalized list of capability identifiers from either a full
  ampersand document or a raw capability map.
  """
  @spec normalize(input()) :: [capability_id()]
  def normalize(input) do
    input
    |> capabilities_from_input!()
    |> Map.keys()
    |> Enum.uniq()
    |> Enum.sort()
  end

  @doc """
  Composes one or more capability maps or full documents into a single
  capability map.

  If the same capability appears multiple times with the same binding, the
  duplicate collapses cleanly. If the same capability appears with different
  bindings, composition fails with a conflict.
  """
  @spec compose([input()]) :: compose_result()
  def compose(inputs) when is_list(inputs) do
    Enum.reduce_while(inputs, {:ok, identity()}, fn input, {:ok, acc} ->
      case capabilities_from_input(input) do
        {:ok, capabilities} ->
          case merge(acc, capabilities) do
            {:ok, merged} -> {:cont, {:ok, merged}}
            {:error, _reason} = error -> {:halt, error}
          end

        {:error, _reason} = error ->
          {:halt, error}
      end
    end)
  end

  @doc """
  Checks whether two declarations are equivalent after normalization.
  """
  @spec aci_equivalent?(input(), input()) :: boolean()
  def aci_equivalent?(left, right) do
    normalize(left) == normalize(right) and bindings_equivalent?(left, right)
  end

  @doc """
  Checks the commutative property for two compatible declarations.
  """
  @spec commutative?(input(), input()) :: boolean()
  def commutative?(left, right) do
    case {compose([left, right]), compose([right, left])} do
      {{:ok, forward}, {:ok, reverse}} -> forward == reverse
      _ -> false
    end
  end

  @doc """
  Checks the associative property for three compatible declarations.
  """
  @spec associative?(input(), input(), input()) :: boolean()
  def associative?(left, middle, right) do
    with {:ok, left_middle} <- compose([left, middle]),
         {:ok, middle_right} <- compose([middle, right]),
         {:ok, grouped_left} <- compose([left_middle, right]),
         {:ok, grouped_right} <- compose([left, middle_right]) do
      grouped_left == grouped_right
    else
      _ -> false
    end
  end

  @doc """
  Checks the idempotent property for a declaration.
  """
  @spec idempotent?(input()) :: boolean()
  def idempotent?(input) do
    case {compose([input]), compose([input, input])} do
      {{:ok, once}, {:ok, twice}} -> once == twice
      _ -> false
    end
  end

  @doc """
  Checks the identity property for a declaration.
  """
  @spec identity?(input()) :: boolean()
  def identity?(input) do
    case {compose([identity(), input]), compose([input, identity()]), compose([input])} do
      {{:ok, left_identity}, {:ok, right_identity}, {:ok, original}} ->
        left_identity == original and right_identity == original

      _ ->
        false
    end
  end

  @doc false
  @spec merge(capability_map(), capability_map()) :: compose_result()
  def merge(left, right) when is_map(left) and is_map(right) do
    Enum.reduce_while(right, {:ok, left}, fn {capability, binding}, {:ok, acc} ->
      case Map.fetch(acc, capability) do
        :error ->
          {:cont, {:ok, Map.put(acc, capability, binding)}}

        {:ok, existing} when existing == binding ->
          {:cont, {:ok, acc}}

        {:ok, existing} ->
          {:halt, {:error, {:conflicting_binding, capability, existing, binding}}}
      end
    end)
  end

  defp bindings_equivalent?(left, right) do
    case {capabilities_from_input(left), capabilities_from_input(right)} do
      {{:ok, left_caps}, {:ok, right_caps}} -> left_caps == right_caps
      _ -> false
    end
  end

  defp capabilities_from_input!(input) do
    case capabilities_from_input(input) do
      {:ok, capabilities} -> capabilities
      {:error, {:invalid_input, reason}} -> raise ArgumentError, "invalid composition input: #{inspect(reason)}"
    end
  end

  defp capabilities_from_input(%{"capabilities" => capabilities}) when is_map(capabilities),
    do: {:ok, capabilities}

  defp capabilities_from_input(%{capabilities: capabilities}) when is_map(capabilities),
    do: {:ok, capabilities}

  defp capabilities_from_input(input) when is_map(input) do
    if capability_map?(input) do
      {:ok, input}
    else
      {:error, {:invalid_input, input}}
    end
  end

  defp capabilities_from_input(input), do: {:error, {:invalid_input, input}}

  defp capability_map?(map) when map_size(map) == 0, do: true

  defp capability_map?(map) do
    Enum.all?(map, fn
      {capability, binding} when is_binary(capability) and is_map(binding) -> true
      _ -> false
    end)
  end
end

File: sdk/python/ampersand_protocol/src/ampersand_protocol/compose.py
"""Capability composition and normalization helpers for ampersand declarations.

This module provides deterministic, set-like behavior for capability maps:

- normalize capability identifiers into a sorted list
- compose one or more capability maps/documents
- detect conflicting bindings
- expose lightweight ACI property checks
"""

from __future__ import annotations

from dataclasses import dataclass
import json
from pathlib import Path
from typing import Any, Dict, List, Mapping, Optional, Sequence, Tuple

CapabilityId = str
CapabilityBinding = Dict[str, Any]
CapabilityMap = Dict[CapabilityId, CapabilityBinding]
Document = Mapping[str, Any]


@dataclass(frozen=True)
class ComposeConflict:
    """Structured conflict information for a capability merge failure."""

    capability: str
    existing: Mapping[str, Any]
    incoming: Mapping[str, Any]

    def to_dict(self) -> Dict[str, Any]:
        return {
            "capability": self.capability,
            "existing": dict(self.existing),
            "incoming": dict(self.incoming),
        }


@dataclass(frozen=True)
class ComposeResult:
    """Normalized success/error wrapper for compose operations."""

    ok: bool
    errors: List[str]
    document: Optional[Dict[str, Any]] = None
    conflict: Optional[ComposeConflict] = None

    def to_dict(self) -> Dict[str, Any]:
        payload: Dict[str, Any] = {"ok": self.ok, "errors": self.errors}
        if self.document is not None:
            payload["document"] = self.document
        if self.conflict is not None:
            payload["conflict"] = self.conflict.to_dict()
        return payload


class ComposeError(ValueError):
    """Raised when composition inputs are invalid or conflicting."""

    def __init__(self, message: str, *, conflict: Optional[ComposeConflict] = None) -> None:
        super().__init__(message)
        self.conflict = conflict


def identity() -> CapabilityMap:
    """Return the identity capability map for composition."""
    return {}


def normalize_capabilities(input_value: Mapping[str, Any]) -> List[str]:
    """Return sorted capability identifiers from a declaration or capability map."""
    capabilities = _capabilities_from_input(input_value)
    return sorted(set(capabilities.keys()))


def compose(inputs: Sequence[Mapping[str, Any]]) -> CapabilityMap:
    """Compose one or more declarations/capability maps into a single capability map.

    Rules:
    - disjoint keys merge directly
    - identical duplicate bindings collapse cleanly
    - conflicting bindings for same key raise ComposeError
    """
    merged: CapabilityMap = {}

    for input_value in inputs:
        incoming = _capabilities_from_input(input_value)
        merged = merge(merged, incoming)

    return merged


def compose_capabilities(inputs: Sequence[Mapping[str, Any]]) -> ComposeResult:
    """
    Compose one or more declarations/capability maps and return a result wrapper.

    Returns:
      ComposeResult(ok=True, document={"capabilities": ..., "capability_count": ...})
      ComposeResult(ok=False, errors=[...], conflict=...)
    """
    try:
        merged = compose(inputs)
    except ComposeError as exc:
        return ComposeResult(
            ok=False,
            errors=[str(exc)],
            conflict=exc.conflict,
        )

    return ComposeResult(
        ok=True,
        errors=[],
        document={
            "capabilities": merged,
            "capability_count": len(merged),
            "normalized": sorted(merged.keys()),
        },
    )


def compose_documents(paths: Sequence[str | Path]) -> ComposeResult:
    """
    Load declaration JSON files, compose their capabilities, and return result wrapper.
    """
    documents: List[Dict[str, Any]] = []

    for raw_path in paths:
        path = Path(raw_path).expanduser().resolve()
        try:
            with path.open("r", encoding="utf-8") as f:
                decoded = json.load(f)
        except OSError as exc:
            return ComposeResult(
                ok=False,
                errors=[f"Unable to read declaration at {path}: {exc}"],
            )
        except json.JSONDecodeError as exc:
            return ComposeResult(
                ok=False,
                errors=[f"Unable to parse declaration at {path}: {exc}"],
            )

        if not isinstance(decoded, dict):
            return ComposeResult(
                ok=False,
                errors=[f"Declaration at {path} must decode to a JSON object"],
            )

        documents.append(decoded)

    return compose_capabilities(documents)


def merge(left: Mapping[str, Any], right: Mapping[str, Any]) -> CapabilityMap:
    """Merge two capability maps with conflict detection."""
    if not isinstance(left, Mapping) or not isinstance(right, Mapping):
        raise ComposeError("left and right must be capability maps")

    merged: CapabilityMap = {str(k): _as_binding(v, context="left map") for k, v in left.items()}

    for capability, incoming_binding_raw in right.items():
        capability_id = _validate_capability_key(capability)
        incoming_binding = _as_binding(incoming_binding_raw, context=f"capability {capability_id}")

        if capability_id not in merged:
            merged[capability_id] = incoming_binding
            continue

        existing_binding = merged[capability_id]
        if _bindings_equal(existing_binding, incoming_binding):
            continue

        conflict = ComposeConflict(
            capability=capability_id,
            existing=existing_binding,
            incoming=incoming_binding,
        )
        raise ComposeError(
            f"conflicting binding for capability {capability_id}",
            conflict=conflict,
        )

    return merged


def aci_equivalent(left: Mapping[str, Any], right: Mapping[str, Any]) -> bool:
    """Check whether two declarations are equivalent under normalization + bindings."""
    try:
        left_caps = _capabilities_from_input(left)
        right_caps = _capabilities_from_input(right)
    except ComposeError:
        return False

    if sorted(left_caps.keys()) != sorted(right_caps.keys()):
        return False

    return _bindings_equal(left_caps, right_caps)


def commutative(left: Mapping[str, Any], right: Mapping[str, Any]) -> bool:
    """Check commutativity for two composition inputs."""
    try:
        return compose([left, right]) == compose([right, left])
    except ComposeError:
        return False


def associative(
    left: Mapping[str, Any],
    middle: Mapping[str, Any],
    right: Mapping[str, Any],
) -> bool:
    """Check associativity for three composition inputs."""
    try:
        grouped_left = compose([compose([left, middle]), right])
        grouped_right = compose([left, compose([middle, right])])
        return grouped_left == grouped_right
    except ComposeError:
        return False


def idempotent(input_value: Mapping[str, Any]) -> bool:
    """Check idempotency for a composition input."""
    try:
        return compose([input_value]) == compose([input_value, input_value])
    except ComposeError:
        return False


def has_identity(input_value: Mapping[str, Any]) -> bool:
    """Check identity law for a composition input."""
    try:
        one = compose([input_value])
        return compose([identity(), input_value]) == one and compose([input_value, identity()]) == one
    except ComposeError:
        return False


def _capabilities_from_input(input_value: Mapping[str, Any]) -> CapabilityMap:
    """Extract and validate a capability map from either:
    - full document: {"capabilities": {...}}
    - raw capability map: {"&memory.graph": {...}, ...}
    """
    if not isinstance(input_value, Mapping):
        raise ComposeError("composition input must be a mapping")

    if "capabilities" in input_value:
        capabilities = input_value.get("capabilities")
        if not isinstance(capabilities, Mapping):
            raise ComposeError("document.capabilities must be a mapping")
        return _validate_capability_map(capabilities)

    return _validate_capability_map(input_value)


def _validate_capability_map(value: Mapping[str, Any]) -> CapabilityMap:
    if not isinstance(value, Mapping):
        raise ComposeError("capability map must be a mapping")

    normalized: CapabilityMap = {}
    for raw_key, raw_binding in value.items():
        capability_id = _validate_capability_key(raw_key)
        normalized[capability_id] = _as_binding(
            raw_binding,
            context=f"capability {capability_id}",
        )
    return normalized


def _validate_capability_key(value: Any) -> str:
    if not isinstance(value, str) or value.strip() == "":
        raise ComposeError(f"capability key must be a non-empty string, got {value!r}")
    if not value.startswith("&"):
        raise ComposeError(f"capability key must start with '&', got {value!r}")
    return value


def _as_binding(value: Any, *, context: str) -> CapabilityBinding:
    if not isinstance(value, Mapping):
        raise ComposeError(f"{context} binding must be a mapping")
    return {str(k): v for k, v in value.items()}


def _bindings_equal(left: Any, right: Any) -> bool:
    return _stable_primitive(left) == _stable_primitive(right)


def _stable_primitive(value: Any) -> Any:
    """Convert nested values into stable comparable primitives."""
    if isinstance(value, Mapping):
        return tuple((str(k), _stable_primitive(v)) for k, v in sorted(value.items(), key=lambda item: str(item[0])))
    if isinstance(value, list):
        return tuple(_stable_primitive(v) for v in value)
    if isinstance(value, tuple):
        return tuple(_stable_primitive(v) for v in value)
    return value


__all__ = [
    "CapabilityBinding",
    "CapabilityMap",
    "ComposeConflict",
    "ComposeError",
    "ComposeResult",
    "aci_equivalent",
    "associative",
    "commutative",
    "compose",
    "compose_capabilities",
    "compose_documents",
    "has_identity",
    "idempotent",
    "identity",
    "merge",
    "normalize_capabilities",
]
